### ISSUE TITLE
Add decideSymbol, decideNat, decideChar, decTypeRep and decT

### ISSUE THREAD

**Issue by phadej** [haskell/core-libraries-committee#98]
```haskell
-- sameSymbol returns Maybe (a :~: b)
decideSymbol :: (KnownSymbol a, KnownSymbol b) =>
              proxy1 a -> proxy2 b -> Either (a :~: b -> Void) (a :~: b)
decideSymbol x y
  | symbolVal x == symbolVal y  = Right (unsafeCoerce Refl)
  | otherwise                   = Left (\Refl -> errorWithoutStackTrace ("decideSymbol: A proof of " ++ show (symbolVal x) ++ " :~: " ++ show (symbolVal y)))
  
-- decideNat and decideChar similarly for Natural and Char.
```

```haskell
-- the eqTypeRep returns Maybe (a :~: b)
decTypeRep :: forall k1 k2 (a :: k1) (b :: k2).
             TypeRep a -> TypeRep b -> Either (a :~~: b -> Void) (a :~~: b)
decTypeRep a b
  | sameTypeRep a b = Right (unsafeCoerce HRefl)
  | otherwise       = Left (\HRefl -> errorWithoutStackTrace ("decTypeRep: got a proof of " ++ show a ++ " :~~: " ++ show b))
  
-- | eqT returns Maybe (a :~: b)
decT :: forall a b. (Typeable a, Typeable b) => Either (a :~: b -> Void) (a :~: b)
decT = ...
```

The `sameSymbol`, `sameNat`, `sameChar` and `eqTypeRep` could be safely&dagger; re-implemented using these new combinators:

```haskell
sameSymbol a b = either (const Nothing) Maybe (decSymbol a b)
```

but it's probably better to not change their implementations, as they are quite simple anyway. 

&dagger; by safely I mean *without relying on any internal implementation details*. Thus if we had decide variants from the beginning, libraries could have defined `Maybe (a :~: b)` variants if they needed them, but defining `decide` variants using `same` functions is relying on (undocumented!) implementation details.

---

This unsafe code already exists in `singletons-base`, and I argue that such unsafe stuff is better to define next to the actual types.

https://hackage.haskell.org/package/singletons-base-3.1.1/docs/src/GHC.TypeLits.Singletons.Internal.html#line-111

```haskell
instance SDecide Symbol where
  (SSym :: Sing n) %~ (SSym :: Sing m)
    | Just r <- sameSymbol (Proxy :: Proxy n) (Proxy :: Proxy m)
    = Proved r
    | otherwise
    = Disproved (\Refl -> error errStr)
    where errStr = "Broken Symbol singletons"
```

and https://hackage.haskell.org/package/singletons-base-3.1.1/docs/src/Data.Singletons.Base.TypeRepTYPE.html#line-93

```haskell
instance SDecide (TYPE rep) where
  tra %~ trb =
    case eqTypeRep tra trb of
      Just HRefl -> Proved Refl
      Nothing    -> Disproved (\Refl -> error "Type.Reflection.eqTypeRep failed")
```

cc @RyanGlScott

**Comment by treeowl** [comment]
I think those should be `decideKnownSymbol`, etc., with a similar `decideTypeable`. Do we have/will we get the actual singleton types in `base`? If so, the functions called `decideSymbol` etc. should take those.

**Comment by RyanGlScott** [comment]
You should be able to implement `decTypeRep` without `unsafeCoerce`:

```hs
decTypeRep :: forall k1 k2 (a :: k1) (b :: k2).
              TypeRep a -> TypeRep b -> Either (a :~~: b -> Void) (a :~~: b)
decTypeRep a b =
  case eqTypeRep a b of
    Just e  -> Right e
    Nothing -> Left (\HRefl -> errorWithoutStackTrace ("decTypeRep: got a proof of " ++ show a ++ " :~~: " ++ show b))
```

I don't have a strong opinion on the naming.

> Do we have/will we get the actual singleton types in `base`?

Yes, those exist now that #85 has been implemented.

**Comment by treeowl** [comment]
> > Do we have/will we get the actual singleton types in `base`?
> 
> Yes, those exist now that #85 has been implemented.

In that case, I think we should definitely have comparison functions that take them, and optionally have the class-based versions.

**Comment by phadej** [comment]
> You should be able to implement `decTypeRep` without `unsafeCoerce`:

Yes, but I still need `error` "which shouldn't ever be reached".

So one might rather implement `eqTypeRep` using `decTypeRep`, not the other way around.

**Comment by treeowl** [comment]
I suggest adding the full set you're considering to the proposal.

**Comment by phadej** [comment]
> I suggest adding the full set you're considering to the proposal.

What do you mean by "the full set"? I'm considering just the four mentioned in the title.

**Comment by phadej** [comment]
I'm also not sure what the GHC.TypeLits interface will look like in GHC-9.6 and happy to postpone moving this proposal forward until it's out.

**Comment by treeowl** [comment]
There are four different kinds: `Type`, `Nat`, `Symbol`, and `Char`. For each of them, I would definitely want a version that accepts explicit singleton/`TypeRep` arguments. It may also be useful to have, for each, a version that takes those implicitly as class constraints. I'm suggesting that you specify whether you want explicit ones, implicit ones, or both, and that you propose names for all four/eight. Your proposal as it stands has a mix of implicit and explicit, with no scheme for differentiating them by name.

**Comment by phadej** [comment]
> I'm suggesting that you specify whether you want explicit ones, implicit ones, or both

I propose only implicit ones for `Nat`, `Symbol` and `Char`, and both for `Typeable/TypeRep`.

The reason for having explicit one for `TypeRep` is because it's hetero-kinded: `:~~: - double tilde.

The reason for not having implicit ones for `Nat`, `Symbol` and `Char` is because I'd rather add a `decideEquality` to `TestEquality` type-class, but that is breaking change and I don't want to dilute this proposal with breaking change concerns.

**Comment by treeowl** [comment]
I see now.

**Comment by Bodigrim** [comment]
I'd probably wait until GHC 9.6 is out and we gain some experience with changes approved in #85 before proceeding with this proposal, but it's proposal's right to trigger a vote if they wish so.

**Comment by phadej** [comment]
This and #85 are orthogonal, so I'd wish to trigger the vote.

**Comment by Bodigrim** [comment]
@phadej could you please raise a draft MR which we can vote on?

**Comment by phadej** [comment]
The MR for this proposal is  blocked on merging https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9112

**Comment by phadej** [comment]
Also this should wait until `heqT` #99 is decided upon, as it would make sense to have `hdecT`.

**Comment by phadej** [comment]
ping @Bodigrim

**Comment by Bodigrim** [comment]
@rae @RyanGlScott could you please opine on this?

**Comment by RyanGlScott** [comment]
I like it. I've always been a bit annoyed that `base` didn't offer primitives like these before, so these would fill a useful gap in the API.

The only thing I might find contentious are the naming conventions: it's a bit strange that some of these functions begin with the prefix `decide`, while others begin with the prefix `dec`. Is there any reason not to use one prefix or the other consistently?

**Comment by phadej** [comment]
@RyanGlScott

To summarise the naming proposed:

```haskell
same*   = instances + proxies -> Maybe (a :~: b)
eq*     = singletons          -> Maybe (a :~: b) -- or `a :~~: b`)

decide* = instance + proxies -> Either ... (a :~: b)
dec*    = singletons         -> Either ... (a :~: b)
```

---

In base we have inconsistencies already:

```haskell
-- instances + proxies
sameNat :: (KnownNat a, KnownNat b) => proxy1 a -> proxy2 b -> Maybe (a :~: b)

-- singletons
eqTypeRep :: forall k1 k2 (a :: k1) (b :: k2). TypeRep a -> TypeRep b -> Maybe (a :~~: b)

-- inconsistent naming, fortunately not exported!
-- I'd come up with own prefix for this.
sameTypeRep :: forall k1 k2 (a :: k1) (b :: k2). TypeRep a -> TypeRep b -> Bool

-- also for, which is also breaks the naming pattern.
eqStableName :: StableName a -> StableName b -> Bool

-- eqT also doesn't fit any scheme, but it's unique so whatever.
eqT :: forall a b. (Typeable a, Typeable b) => Maybe (a :~: b) -- no proxy params!
```

So the names I added are 

```haskell
-- decide: instances + proxies
decideSymbol :: (KnownSymbol a, KnownSymbol b) => proxy1 a -> proxy2 b -> Either (a :~: b -> Void) (a :~: b)

-- dec: singletons
decTypeRep :: forall k1 k2 (a :: k1) (b :: k2). TypeRep a -> TypeRep b -> Either (a :~~: b -> Void) (a :~~: b)
```

One could imagine adding

```haskell
-- singletons
-- but this is the same as testEquality
eqNat :: SNat -> SNat b -> Maybe (a :~: b)

-- instances + proxies
-- but this clashes with internal name and
-- is just eqT with proxy arguments
sameTypeRep :: (Typeable a, Typeable b) => proxy1 a -> proxy2 b -> Maybe (a :~: b)
```

**Comment by RyanGlScott** [comment]
Ah, that is a convincing reason to have two separate naming conventions. I also like that this would make it possible to add functions like `decSymbol :: SSymbol a -> SSymbol b -> Either (a :~: b -> Void) (a :~: b)` later if we wanted them.

+1 from me.

**Comment by treeowl** [comment]
Why do these functions take proxies? Their return types determine their type arguments.

**Comment by phadej** [comment]
The `same` variants take (from their original addition in 2014). I don't think it's worth breaking the pattern at this point.

**Comment by Bodigrim** [comment]
Dear CLC members, before we move to a vote, please take a look at the MR https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9524. If you are not content with it as is, what kind of changes could sway you to vote in favour?
@tomjaguarpaw @chessai @cgibbard @emilypi @mixphix

**Comment by cgibbard** [comment]
Seems good, I imagine if you're profitably using such proofs, you'd appreciate having the refutations sometimes. Personally I try to stay away from singletons and fancy type level data, that's sort of right where I draw the line right now with Haskell's type system, but if I was using that stuff, I can imagine this change being welcome.

It might be worth checking that the code for eqTypeRep ends up being the same as it used to, for anyone who might be concerned with the performance of this stuff. I can imagine that after inlining and a bit of case specialization it probably does end up simplifying to the same thing?

**Comment by phadej** [comment]
@cgibbard 

> It might be worth checking that the code for eqTypeRep

I did. Check collapsed thread with this comment https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9524#note_468857

**Comment by phadej** [comment]
@Bodigrim could you trigger a vote?

**Comment by Bodigrim** [comment]
Dear CLC members, let's vote on the proposal to add `decT` / `hdecT` to `Data.Typeable`, `decTypeRep` to `Type.Reflection`, `decideNat` to `Data.TypeNats`, and finally `decideSymbol` / `decideChar` to `Data.TypeLits`. The details can be found in https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9524.
@tomjaguarpaw @chessai @cgibbard @emilypi @mixphix 

----

+1 from me. This is very far away from my comfort zone, but I trust @RyanGlScott's [opinion](https://github.com/haskell/core-libraries-committee/issues/98#issuecomment-1356842905).

**Comment by cgibbard** [comment]
+1

**Comment by tomjaguarpaw** [comment]
+1

----

Very nice, thanks!

**Comment by chessai** [comment]
+1

Nice, thanks

**Comment by Bodigrim** [comment]
Thanks all, 4 votes in favour are enough to approve.

**Comment by parsonsmatt** [comment]
I would prefer to avoid `dec*` as a name prefix - `dec` often expands to `declaration`, where `decidable` is much less common IME. I wouldn't le that override a prior decision though.

`data Decision a` seems like a good addition - a more specific type can carry documentation and suggestive labels that greatly aid in legibility. And writing `Decision a` instead of `Either (a -> Void) a` feels clearly better to me.

+1

**Comment by hasufell** [comment]
> and the return type would look nicer

Well, for me that doesn't sound too interesting to change the type signature yet.

> There are some instances Decision could be, e.g. Invariant functor.

More interesting. Do we know what is the demand?

**Comment by phadej** [comment]
> More interesting. Do we know what is the demand?

No. But two people (me and Ryan) defined it independently, and the similar type exist in [Agda stdlib](https://agda.github.io/agda-stdlib/Relation.Nullary.Decidable.Core.html#1476) (wherefrom I got inspiration for my `Dec`) for the same purpose. It has `map` function (that would be `invmap` from `Invariant` class)

**Comment by phadej** [comment]
> but I hope the authors are confident that this is the right API

Neither choice

```haskell
decideSymbol :: (KnownSymbol a, KnownSymbol b) => proxy1 a -> proxy2 b -> Either (a :~: b -> Void) (a :~: b)
```

or

```haskell
decideSymbol :: (KnownSymbol a, KnownSymbol b) => proxy1 a -> proxy2 b -> Decision (a :~: b)
```

is more correct than other

---

Specifically, I'm asking CLC to make the decision between the two.

- Either you bless that `Either` is ok, and we won't need `Decision` any time soon (or *you* argue that it should be provided in libraries, like `singletons`)
- or you help design `Decision` (e.g. names, in which module to put in, etc.), 
- or you softly accept the possible breaking change in the future, if we now do `Either` but good reasons for `Decision` become needed later
- or we just drop this proposal, as indeed, this is an addition which can be done in a libraries, and in fact is in `singletons`.

**Comment by hasufell** [comment]
> is more correct than other

That was not my question.

**Comment by phadej** [comment]
@hasufell could you clarify your question then?

**Comment by mixphix** [comment]
I'm -1 on `Decision` being a part of the public-facing API of `base` as a part of this proposal.

If the GHC devs want to use one single name for this type, so be it: make `GHC.Decision` and see how far use spreads, then raise a *separate proposal* to expose `Data.Decision`. I don't see how this is relevant at all to the content of the proposal being discussed.

**Comment by Bodigrim** [comment]
@phadej please land the approved proposal into `base` first. If there is interest in `data Decision`, please raise a separate proposal, there is plenty of time to discuss before GHC 9.8.

**Comment by phadej** [comment]
@Bodigrim then changing these functions to use `data Decision` as return type would be a breaking change. I want to get this right from get go.

**Comment by Bodigrim** [comment]
It won't be a breaking change until GHC 9.8 is released.

**Comment by phadej** [comment]
It won't, but I have no energy to follow up in a timely manner.

**Comment by Bodigrim** [comment]
There've been plenty of time to discuss the proposal and any interested party could have suggested `data Decision a` during that timeframe. This didn't happened, and a vote has been triggered, and the proposal has been approved as is. 

I don't think `Decision a` makes a material difference: it is slightly fancier, but also a new ad-hoc concept entailing all sorts of consequences (e. g., getting instances right). I personally don't have energy to spend on another month of lukewarm ramblings around it. We cannot put proposals on hold just and reopen the floor each time someone comes up with a suggestion (especially at the very last moment), this is a certain way to do nothing.

**Comment by phadej** [comment]
@Bodigrim and GHC dev did in their review, please answer them

>  I'm just wondering whether we really want to use Either, and not a custom datatype like SDecide for this? 

https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9524#note_485032

I'm happy with either outcome, and I'm neutral about having `Decision`.

**Comment by Bodigrim** [comment]
I've answered, but AFAIU @sheaf has approved the MR and - unless **you personally** want to pursue `data Decision a` - you can go ahead with a merge.

**Comment by phadej** [comment]
> **you personally**

I don't

**Comment by hasufell** [comment]
> @hasufell could you clarify your question then?

I was asking about your confidence level of the newly proposed API. My impression is that it's not very high and there's no strong use case for instances, so we have no idea how future proof this is. As such, I also view this conservatively.

**Comment by Icelandjack** [comment]
> There are some instances Decision could be, e.g. [Invariant functor](https://hackage.haskell.org/package/invariant-0.6.1/docs/Data-Functor-Invariant.html#t:Invariant).

There is an option of splitting the parameters to `Decision`, and then defining `type Decision a = Decision_ a a` rather than using the rather useless Invariant we get a Profunctor instance, and maybe some others from that hierarchy.

```haskell
type Decision_ :: Type -> Type -> Type
data Decision_ a b
  = Proved b
  | Disproved (Refuted a)

instance Profunctor Decision_ where
  dimap :: (a' -> a) -> (b -> b') -> (Decision_ a b -> Decision_ a' b')
  dimap pre post = \case
    Proved    yes -> Proved    (post yes)
    Disproved no  -> Disproved (no . pre)
```

**Comment by mixphix** [comment]
Again, unless you personally @Icelandjack are willing to raise `data Decision` as a separate proposal I see no use further discussing the matter in this thread about adding certain functions to type equalities.

**Comment by chshersh** [comment]
I'm trying to summarise the state of this proposal as part of my volunteering effort to track the progress of all `approved` CLC proposals.

| Field | Value |
|---|---|
| **Authors**            |  @phadej   |
| **Status**             |  merged  |
| **`base` version**     | `4.19.0.0`  |
| **Merge Request (MR)** | https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9524 |
| **Blocked by**         | nothing |
| **CHANGELOG entry**    | present  |
| **Migration guide**    |  not needed |

[changelog]: https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/changelog.md

Please, let me know if you find any mistakes 🙂
